### PR TITLE
Improve JSON parsing to handle more edge cases.

### PR DESCRIPTION
**Pull request by grp**
Using a JSON parser test suite, refine number and key parsing to match
the JSON specification. Remaining issues are related to Unicode support,
escape sequences, and trailing commas.

The last of those is intentionally non-conforming for compatibility.

### CODE REVIEW

File: Libraries/plist/Sources/Format/ASCIIPListLexer.cpp [hunk 1]
```diff
@@ -216,7 +215,7 @@ ASCIIPListLexerReadData(ASCIIPListLexer *lexer)
 static int
 ASCIIPListLexerReadString(ASCIIPListLexer *lexer)
 {
-    if (lexer->pointer[0] == '\'') {
+    if (lexer->pointer[0] == '\'' && lexer->style == kASCIIPListLexerStyleASCII) {
         return ASCIIPListLexerReadSingleQuotedString(lexer);
     } else if (lexer->pointer[0] == '\"') {
```
Comment (Ktwu) [inline Libraries/plist/Sources/Format/ASCIIPListLexer.cpp]: Why doesn't the '\"' check get the style check too?

File: Libraries/plist/Sources/Format/ASCIIPListLexer.cpp [hunk 2]
```diff
@@ -226,27 +225,39 @@ ASCIIPListLexerReadString(ASCIIPListLexer *lexer)
...
         p++;
+    }
 
-    if (!isdigit(*p))
+    if (!isdigit(*p)) {
         return kASCIIPListLexerInvalidToken;
+    }
+    bool zero = (*p == '0');
+    while (isdigit(*p)) {
+        /* Numbers cannot start with zero. */
```
Comment (Ktwu) [inline Libraries/plist/Sources/Format/ASCIIPListLexer.cpp]: So something like "00000" is valid?

Comment (Ktwu) [inline Libraries/plist/Sources/Format/ASCIIPListLexer.cpp]: What about 0000.0000?

File: Libraries/plist/Tests/Format/test_JSON.cpp [hunk 3]
```diff
@@ -98,3 +98,38 @@ TEST(JSON, Empty)
...
 
+TEST(JSON, SingleQuotes)
+{
+    /* Single quotes are not allowed in JSON. */
+    auto contents = Contents("{ 'one': 1 }");
+    auto deserialize = JSON::Deserialize(contents, JSON::Create());
+    EXPECT_EQ(deserialize.first, nullptr);
+}
+
+TEST(JSON, Number)
```
Comment (Ktwu) [inline Libraries/plist/Tests/Format/test_JSON.cpp]: What about testing that the right number is decoded?

Comment (Ktwu) [inline Libraries/plist/Tests/Format/test_JSON.cpp]: What about a test for -1?

File: Libraries/plist/Sources/Format/ASCIIPListLexer.cpp (hunk 2 above)
Comment (grp) [inline Libraries/plist/Sources/Format/ASCIIPListLexer.cpp]: Yeah, that is valid.

File: Libraries/plist/Sources/Format/ASCIIPListLexer.cpp (hunk 1 above)
Comment (grp) [inline Libraries/plist/Sources/Format/ASCIIPListLexer.cpp]: JSON keys are double-quoted:

    { "key" : "value" }

But not single-quoted:

     { 'key' : 'value' }

ASCII property lists support both.

File: Libraries/plist/Tests/Format/test_JSON.cpp (hunk 3 above)
Comment (grp) [inline Libraries/plist/Tests/Format/test_JSON.cpp]: good idea, will add -1. and test for the numbers

File: Libraries/plist/Sources/Format/ASCIIPListLexer.cpp (hunk 1 above)
Comment (grp) [inline Libraries/plist/Sources/Format/ASCIIPListLexer.cpp]: Yep, agree that's better.

File: Libraries/plist/Tests/Format/test_JSON.cpp (hunk 3 above)
Comment (grp) [inline Libraries/plist/Tests/Format/test_JSON.cpp]: Turns out I can't test the numbers because this is testing the cases where decoding fails so there's no number to check. :P

File: Libraries/plist/Sources/Format/ASCIIPListLexer.cpp (hunk 1 above)
Comment (grp) [inline Libraries/plist/Sources/Format/ASCIIPListLexer.cpp]: Done.
